File: core/src/ru/myitschool/dcrawler/ui/layers/DynamicTileLayer.java
package ru.myitschool.dcrawler.ui.layers;

import com.badlogic.gdx.maps.tiled.TiledMapTileLayer;
import ru.myitschool.dcrawler.dungeon.DungeonCell;


/**
 * Created by Voyager on 20.04.2017.
 */
public class DynamicTileLayer extends TiledMapTileLayer {

    protected int width;
    protected int height;
    private int startX;
    private int startY;

    private float tileWidth;
    private float tileHeight;

    protected DungeonCell[][] cells;

    public DynamicTileLayer(int width, int height, int tileWidth, int tileHeight) {
        super(width, height, tileWidth, tileHeight);
        startX = 0;
        startY = 0;
        this.width = width;
        this.height = height;
        this.tileWidth = tileWidth;
        this.tileHeight = tileHeight;
        this.cells = new DungeonCell[width][height];
    }

    @Override
    public int getWidth() {
        return width;
    }

    @Override
    public int getHeight() {
        return height;
    }

    @Override
    public float getTileWidth() {
        return tileWidth;
    }

    @Override
    public float getTileHeight() {
        return tileHeight;
    }

    @Override
    public DungeonCell getCell(int x, int y) {
        if (x < cells.length && x >= 0 && y < cells[0].length && y >= 0){
            return cells[x][y];
        }
        return null;
    }

    @Override /** Better don't use this method*/
    public void setCell(int x, int y, Cell cell) {
        setCell(x, y, (DungeonCell) cell);
    }

    public void setCell(int x, int y, DungeonCell cell){
        if (x < cells.length && x >= 0 && y < cells[0].length && y >= 0) {
            cells[x][y] = cell;
            cell.setX(x);
            cell.setY(y);
        }
    }

    public int getStartX() {
        return startX;
    }

    public int getStartY() {
        return startY;
    }

    public void addLeft(int count){
        width += count;
        startX += count;
        DungeonCell[][] newCells = new DungeonCell[width][height];
        for (int i = count; i < width; i++){
            for (int j = 0; j < height; j++) {
                newCells[i][j] = cells[i - count][j];
                if (newCells[i][j] != null) {
                    newCells[i][j].setX(i);
                }
            }
        }
        cells = newCells;
    }

    public void addRight(int count){
        DungeonCell[][] newCells = new DungeonCell[width + count][height];
        for (int i = 0; i < width; i++){
            for (int j = 0; j < height; j++) {
                newCells[i][j] = cells[i][j];
            }
        }
        width += count;
        cells = newCells;
    }

    public void addDown(int count){
        DungeonCell[][] newCells = new DungeonCell[width][height + count];
        for (int i = 0; i < width; i++){
            for (int j = 0; j < height; j++) {
                newCells[i][j] = cells[i][j];
            }
        }
        height += count;
        cells = newCells;
    }

    public void addUp(int count){
        height += count;
        startY +=  count;
        DungeonCell[][] newCells = new DungeonCell[width][height];
        for (int i = 0; i < width; i++){
            for (int j = count; j < height; j++) {
                newCells[i][j] = cells[i][j - count];
                if (newCells[i][j] != null) {
                    newCells[i][j].setY(j);
                }
            }
        }
        cells = newCells;
    }

    public DungeonCell[][] getCells() {
        return cells;
    }

    public void setCells(DungeonCell[][] cells) {
        this.cells = cells;
    }

    public void clearLayer(){
        setCells(new DungeonCell[width][height]);
    }
}

File: core/src/com/wolg_vlad/dcrawler/ui/layers/DynamicTileLayer.kt
package com.wolg_vlad.dcrawler.ui.layers

import com.badlogic.gdx.maps.tiled.TiledMapTileLayer
import com.wolg_vlad.dcrawler.dungeon.DungeonCell

/**
 * Created by Voyager on 20.04.2017.
 */
open class DynamicTileLayer(var layerWidth: Int, var layerHeight: Int, tileWidth: Int, tileHeight: Int) : TiledMapTileLayer(layerWidth, layerHeight, tileWidth, tileHeight) {
    var startX = 0
        private set
    var startY = 0
        private set
    private val tileWidth: Float = tileWidth.toFloat()
    private val tileHeight: Float = tileHeight.toFloat()
    var cells: Array<Array<DungeonCell?>>

    override fun getHeight(): Int {
        return layerHeight
    }

    override fun getWidth(): Int {
        return layerWidth
    }

    override fun getTileWidth(): Float {
        return tileWidth
    }

    override fun getTileHeight(): Float {
        return tileHeight
    }

    override fun getCell(x: Int, y: Int): DungeonCell? {
        return if (x < cells.size && x >= 0 && y < cells[0].size && y >= 0) {
            cells[x][y]
        } else null
    }

    /** Better don't use this method */
    override fun setCell(x: Int, y: Int, cell: Cell) {
        setCell(x, y, cell as DungeonCell)
    }

    fun setCell(x: Int, y: Int, cell: DungeonCell) {
        if (x < cells.size && x >= 0 && y < cells[0].size && y >= 0) {
            cells[x][y] = cell
            cell.x = x
            cell.y = y
        }
    }

    fun addLeft(count: Int) {
        layerWidth += count
        startX += count
        val newCells = Array<Array<DungeonCell?>>(layerWidth) { i ->
            return@Array Array(layerHeight) { j ->
                return@Array if (i >= count) { cells[i - count][j]?.x = i
                    cells[i - count][j] }
                else null
            }
        }
        cells = newCells
    }

    fun addRight(count: Int) {
        layerWidth += count
        val newCells = Array<Array<DungeonCell?>>(layerWidth) { i ->
            return@Array Array(layerHeight) { j ->
                return@Array if (i < layerWidth - count) cells[i][j]
                else null
            }
        }
        cells = newCells
    }

    fun addDown(count: Int) {
        layerHeight += count
        val newCells = Array<Array<DungeonCell?>>(layerWidth) { i ->
            return@Array Array(layerHeight) { j ->
                return@Array if (j < layerHeight - count) cells[i][j]
                else null
            }
        }
        cells = newCells
    }

    fun addUp(count: Int) {
        layerHeight += count
        startY += count
        val newCells = Array<Array<DungeonCell?>>(layerWidth) { i ->
            return@Array Array(layerHeight) { j ->
                return@Array if (j >= count) { cells[i][j - count]?.y = j
                    cells[i][j - count] }
                else null
            }
        }
        cells = newCells
    }

    fun clearLayer() {
        cells = Array<Array<DungeonCell?>>(layerWidth) { arrayOfNulls(layerHeight) }
    }

    init {
        this.cells = Array<Array<DungeonCell?>>(layerWidth) { arrayOfNulls(layerHeight) }
    }
}